### PR TITLE
Add peer visibility to MultiplayerSynchronizer.

### LINKED ISSUE #3904: Add interest management support to the high-level multiplayer API
### Describe the project you are working on

And client-server multiplayer game with an authoritative server and large worlds, whether 2D or 3D.

### Describe the problem or limitation you are having in your project

# Problem Statement

When working with large worlds in client-server online multiplayer games with an authoritative server, the game world will have a lot going on all the time. Whether it's other players running around and doing various actions, NPC/Enemy AI, World Events, etc... it would be unreasonable to expect a player's computer to be able to handle processing and rendering everything in the game world at once. 

Even if you break your world up into various instanced "zones," like many games choose to do, depending on the particular zone, it's still likely going to be too much data to process. Level streaming helps out on the rendering side of things, but it does little to assist in the processing department. Additionally, most of the stuff going on in the world or instanced zone isn't even going to be of interest (heh) to you since it's so far away that you likely can't even see it even if you wanted to. 

The solution to this problem that is pretty much ubiquitous in large online worlds these days is Interest Management.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

# Interest Management

Interest Management can take a couple different forms, but ultimately, it's just reducing the number of things you send netowrk updates for to each player. A particular player will only recieve updates from game actors that are in their "area of interest" (a.k.a. "zone of interest",  often abbreviated "AoI" or "ZoI"). In fact, game actors won't even be spawned into the player's client's game world until they are within their AoI and will be promply despawned from the player's local game world once they leave the AoI again.

# Types of Interest Management

There are actually several different implementation approaches out there for interest management, however by and far the two most common are distance-based, and region-based. I'll be focusing on these two for the rest of this proposal. If you're interested in reading more about all of the various implementations out there for interest management, [this paper](https://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.135.565&rep=rep1&type=pdf) does a great job of explaining them and each of their pros and cons.

## Euclidean Distance

This is a popular one because we.... it's dead simple.  A player will subscribe to server updates from any actor that falls within a set radius around them. In this implementation, this radius is often referred to as the player's "aura." While this is dead simple to implement, and computing the distance between two points in space is extremely fast, it has the major problem that it doesn't scale well at all. The more players that are in the game world or instance zone, the more comparisons need to be made every frame. It's for this reason, that the next approach is by and far the most common.

## Grid-based

In this implementation, the world--or zone--is divided into a 2D grid of regions. As actors move about the world, some component (we'll call it the InterestManager here for lack of a  better term) is responsible for keeping track of of what actors are in what regions. Any time the server sends an update by an actor, this InterestManager component is consulted to determine which specific peers should receive the update, and it only sends the update to those peers. Generally, this will not only include the other actors in the same region as that actor, but also those in the immediately adjacent regions as well (or sometimes a higher adjacency is used. The specific region size and adjacency are parameters that are unique to each game and require careful tuning to get right). Likewise, when the InterestManager notices that an actor changes regions, it will recalculate its interested parties, and spawn the actor in the game clients of newly interested players, and despawn it from those who are no longer interested.

Unlike the distance approach above, this one scales incredibly well, but it is much more complicated to implement. The biggest downside for this approach however, is that a simple 2D grid is a rather poor approximation of interest, so for games that require higher degrees of precision, other approaches might be considered instead. 

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

# Proposed Implementation

**UPDATE: @faless created a much more elegantly generic API that better fits core [here](https://github.com/godotengine/godot-proposals/issues/3904#issuecomment-1047769140).  His proposed implementation supersedes the one documented here, but I left this one for context.**

<details><summary><i>Original Implementation</i></summary><p>

First, a common interface will be created that all interest management implementations can inherit from.

```cpp
class InterestManager : public Node {


    // Updates a node's position with the manager. This should be called for every actor
    // on the server any time they move. If appripriate get_multiplayer().get_replicator()._send_spawn()
    // and get_multiplayer().get_replicator()._send_despawn() should be called to (de)spawn the nodes on 
    // specific peers that gain or lose interest.
    //
    // We don't use a Node2D or Node3D directly
    // here so that we can have a single interface that works for both.
    //
    // Note that we are always using a Vector2 here. This is because in most 3D games, there isn't much verticality, so we only
    // really care about 2D positioning. I think this should be fine, even if a game does care about verticality, but I'm 
    // not entirely certain.
    virtual void update_node_position(Node *node, Vector2 position) = 0;
};
```

With this simple interface, we can implement different types of interest management and users can either put them directly in their scene or even add one as a Singleton. Additionally, multiple implementations could theoretically be used together as well. Whetehr or not that's actually useful in practice however, I'm not so sure.

Now let's take a looks at some pseudocode for a grid-based implementation.

```cpp
// grid_interest.h

// Implements a region-based solution to interest management described above.
// This is a TOOL class so that it can render a grid in the editor viewport when
// selected in the scene tree. See draw_grid below.
class GridInterestManager : public InterestManager {
protected:

    // InterestGrid here is just a basic class that is used here to abstract the complexities of managing the region grid.
    InterestGrid grid;

public:
    virtual void update_node_position(network_id : int, Node *node, Vector2 position);

    // Updates the overall size of the grid. Should be used sparingly as this is expensive due to
    // needing to recalculate the regions and who is in what region. Generally only done at _ready.
    //
    // exposed in editor inspector.
    void set_grid_bounds(Vector2 grid_bounds) { grid.set_bounds(grid_bounds); }

    // exposed in editor inspector
    Vector2 get_grid_bounds() { return grid.get_bounds(); }

    // Updates the size of each region (cell) in the grid. Similar to set_grid_bounds, this will cause
    // the grid to recalculate everything so it's very expensive and should be used sparingly. Again,
    // only really necessary at _ready.
    //
    // exposed in editor inspector
    void set_region_size(Vector2 region_size) { grid.set_region_size(region_size); }

    // exposed in editor inspector
    Vector2 get_region_size() { return grid.region_size(); }

    // AreaOfInterest is just an aggregation of 9 regions (a center region and the 8 adjacent ones). All areas of 
    // interest for a grid are calculated when the grid is constructed (or reconstructed) and cached in memory, as 
    // well as each individual region. Additionally, a hashmap of node instance id to AreaOfInterest* is maintained 
    // so that this lookup can be O(1).
    AreaOfInterest* get_aoi_for_node(Node *node);

    // I don't know if this function signature is correct, but the idea here is that whenever this node
    // is selected from the scene tree, it will render the configured grid in the editor viewport.
    void draw_grid(Control *overlay);
};


// grid_interest.cpp
void GridInterestManager::update_node_position(network_id : int, Node *node, Vector2 position) {
    // This will calculate the AreaOfInterest for the node and if it's changed, add the node to
    // the new AoI's list and remove it from the old one's. For each AoI, a list of changes
    // is maintained, which is useful later.
    //
    // Also, if position hasn't changed, this is effectively a no-op.
    grid.update(node, position);
};

void GridInterestManager::remove_node(Node *node) {
    grid.remove_node(node);
};
```

This handles keeping the region list up to date with who's in what region, but now we need to handle those changes somehow. Doing this all in one big for loop obviously isn't great as it will end up bogging the framerate down at runtime. Instead, we will flip the problem on its head and use a second new Node type.

```cpp

// interest_area.h

// InterestArea keeps track of the AoI for a spefic actor.
class InterestArea2D : public Node {

public:
    // These 3 are exposed in the inspector via their paths. actor will default to ".." for convenience.
    Node2D* actor;
    GridInterestManager* interest_manager;
    int network_id;

protected:
    void _spawn_interests(Region *region);
    void _despawn_interests(Region *region);
};

// interest_area.cpp

// Yea, yea. I know that in the cpp code, _notifaction is used directly, but I'm too lazy to 
// write the glue for that here, so this is what you get. Basically, this is just called once per frame.
void InterestArea::_process(float delta) {

    interest_manager->update_node_position(network_id, actor, actor->get_global_position());

    // We don't want to free this when we are done because it's a pointer to data managed by GridInterestManager.
    AreaOfInterest* aoi = interest_manager->get_aoi_for_node(actor);
    _despawn_interests(aoi);
    _spawn_interests(aoi);
    // This simply clears the dirty interest lists so that next time we process them, they only have new stuff.
    region->clean();
}

void InterestArea::_spawn_interests(AreaOfInterest *aoi) {
    for (int j = 0; j < aoi->gained_interest->size(); i++) {
        // Interest is just a pair of node and its network id.
        Interest interest = aoi->gained_interest[1];
        const ObjectID oid = actor->get_instance_id();
        // These two method actually aren't currently exposed, and will need to be so in order for this to work. Alternatively 
        // (and honestly more preferred imo) is th have an exposed `spawn` method that can derive trhe spawner argument from
        // the passed in Node itself.
        MultiplayerSpawner *spawner = get_multiplayer_api()->get_replicator().get_spawner(oid)
        get_multiplayer_api()->get_replicator()->_send_spawn(interest.node, interest.network_id);
    }
}

void InterestArea::_despawn_interests(AreaOfInterest *aoi) {
    for (int j = 0; j < aoi->lost_interest->size(); i++) {
        Interest interest = aoi->lost_interest[1];
        // This method will need exposed for this to work.
        get_multiplayer_api()->get_replicator()->_send_despawn(node, interest.network_id);
    }
}

```

There should exist one InterestArea node for each actor that exists in the world, but there should only be one GridInterestManager for an entire "zone." (Though there doesn't strictly _need_ to be, as mentioned above. Maybe users will get creative on uses here.)

Doing the spawning/despawning in this way should be pretty quick, even in super dense areas, as the number of poeple entering and leaving a AoI should be pretty small.

</p></details>

### LINKED ISSUE #61711: MultiplayerSynchronizer server crash in alpha9+
### Godot version

eaffb88558b6ef2a4c20ad7d1da94e3d1d4c0f4a

### System information

Fedora Silverblue 36, Ryzen 2400G

### Issue description

When running the [bomber](https://github.com/godotengine/godot-demo-projects/tree/4.0-dev/networking/multiplayer_bomber) example, the server instance crashes after starting the game. The lobby itself works fine so basic communications works. This problem does not exist with alpha8.

I have the same issue in my project, and removing `MultiplayerSynchronizer` makes it go away.

```
Godot Engine v4.0.alpha.calinou.eaffb8855 - https://godotengine.org
Vulkan API 1.2.0 - Using Vulkan Device #0: AMD - AMD RADV RAVEN
 

================================================================
handle_crash: Program crashed with signal 11
Engine version: Godot Engine v4.0.alpha.calinou (eaffb88558b6ef2a4c20ad7d1da94e3d1d4c0f4a)
Dumping the backtrace. Please include this when reporting the bug on https://github.com/godotengine/godot/issues
[1] /lib64/libc.so.6(+0x3ea70) [0x7fe5ada24a70] ()
[2] List<NodePath, DefaultAllocator>::front() const ()
[3] List<NodePath, DefaultAllocator>::List(List<NodePath, DefaultAllocator> const&) ()
[4] SceneReplicationInterface::_send_sync(int, unsigned long) ()
[5] SceneReplicationInterface::on_network_process() ()
[6] MultiplayerAPI::poll() ()
[7] SceneTree::process(double) ()
[8] Main::iteration() ()
[9] OS_LinuxBSD::run() ()
[10] /var/home/jgillich/Downloads/godot(main+0x168) [0x556a70394a4d] ()
[11] /lib64/libc.so.6(+0x29550) [0x7fe5ada0f550] ()
[12] /lib64/libc.so.6(__libc_start_main+0x89) [0x7fe5ada0f609] ()
[13] /var/home/jgillich/Downloads/godot(_start+0x2a) [0x556a7039482a] ()
-- END OF BACKTRACE --
================================================================
Aborted (core dumped)

```

### Steps to reproduce

1. Get the bomber example
2. Run two instances and start a game

### Minimal reproduction project

_No response_

### LINKED ISSUE #62320: MultiplayerSpawner not working when instantiated on client after the host
### Godot version

v4.0.alpha10.official [4bbe7f0b9]

### System information

Windows 10

### Issue description

I've been getting some very buggy behavior with the MultiplayerSpawner when instantiating it

If you're connected to a server as a client, and instantiate a scene containing a MultiplayerSpawner AFTER the host has already instantiated that same scene & the host has already instantiated scenes for the MultiplayerSpawner to spawn, the MultiplayerSpawner just won't work on the client

The following error will be spit out by the client:
error

### Steps to reproduce

Simply load a scene containing a MultiplayerSpawner as a client after the host has already loaded that same scene & has instantiated scenes for the MultiplayerSpawner to spawn

### Minimal reproduction project

[MultiplayerSpawner bug.zip](https://github.com/godotengine/godot/files/8960343/MultiplayerSpawner.bug.zip)

### PR DESCRIPTION
**Pull request by Faless**
TL;DR;
---

Sets visibility for synchronized objects: `$MultiplayerSynchronizer.set_visibility_for(peer_id, visible)`

**or**

```gdscript
extends MultiplayerSynchronizer

func _enter_tree():
	add_visibility_filter(self._is_visible_to)

func _is_visible_to(peer) -> bool:
	return false # Hidden
```

The visibility is updated on every frame (default), physics frame, or manually.
You can further notify visibility changes by calling `$MultiplayerSynchronizer.update_visibility(peer:=0)` (where `0` means for all peers).

This also allows players to join mid-game (with some quirks).

Description
--

`MultiplayerSynchronizer`s can now be configured to limit their visibility to a subset of the connected peers, if the synchronized node was spawned by a `MultiplayerSpawner` (either automatically or via custom spawn) the given node will also be de-spawned remotely.

The replication system doesn't have the logic to handle sub-spawn directly, but it is possible to handle them appropriately by manually updating the visibility of the parent before changing the one of the nested spawns via the "update_visibility" function.

The visibility of each `MultiplayerSynchronizer` can specified by overriding `MultiplayerSynchronizer._is_visible_to(peer)`.

To further optimize the network code, visibility can be configured to be automatically updated during idle or physics, or set to always require manual update (`update_visibility` function).

See attached project:

limited_visibility

~[proj_mp_interest.zip](https://github.com/godotengine/godot/files/9095533/proj_mp_interest.zip)~

~*Edit: updated* [proj_mp_interest2.zip](https://github.com/godotengine/godot/files/9106215/proj_mp_interest2.zip)~

*Edit: updated again* [proj_mp_interest3.zip](https://github.com/godotengine/godot/files/9113981/proj_mp_interest3.zip)

I also updated the tests in #62870 to work correctly with one notable hack (delaying world spawning players by one frame, so spawns get replicated in the correct order).

[test8.zip](https://github.com/godotengine/godot/files/9095584/test8.zip)

Note: this PR also adds basic support for **peers joining mid-game** (see demo above) although some more work is likely needed to improve cleanup, and potentially support single-player to multi-player transition.

Note for reviewers:

- MultiplayerSpawner `auto_spawn` property is removed, simply auto spawn whatever is in the "auto spawn list", if `_spawn_custom` is not overridden.
- I don't like having to change MultiplayerAPI, but that's needed for this PR. We should probably do more work and clean it up simplifying the interfaces and modularizing the default implementation as suggested in #62870, but I wanted to keep this PR light.
- There is still some work needed to cleanup the editor plugins, and make sure it shows a warning when the root paths are missing or unreachable for easier debugging.

Fixes https://github.com/godotengine/godot/issues/61711
Fixes https://github.com/godotengine/godot/issues/62320
Closes https://github.com/godotengine/godot-proposals/issues/3904

### CODE REVIEW

**Comment by jgillich** [comment]
AFAIK the most common approach for this is going to be a proximity-based visibility implementation managed by the server. But since authority over MultiplayerSynchronizers is usually granted to the client, the server wouldn't be allowed to manage it, correct? 

The peer ID alone is rarely going to be enough to determine visibility so the `_is_visible_to` method has no use in this case, we would have to manage visibility using RPC instead.

**Comment by reduz** [comment]
@jgillich AFAIK this logic only works server side, client has no authority. I am not entirely sure why you think the peer is not enough, you can obtain the node from it. Maybe one suggestion we could make to @Faless is to allow peers to register metadata (a Variant), so when you implement this function you can pass a pointer or something for other peers to get.

**Comment by reduz** [comment]
@Faless After much though, I still think manually specifying visibility between pairs of peers instead of a callback is likely the way to go instead of this. Callback is OK for a simple game or for extra optional checks.

The reason is that if you have 100 or 1000 moving peers, you probably want to write your visibility code _outside_  your networking layer, then compute the changes between pairs and submit them to the networking layer. 

With the current approach, submitting the changes is quite inefficient because instead of doing.
```GDScript
peer_a.set_peer_visible(peer_b,false)
peer_b.set_peer_visible(peer_a,false)
```
you have to do

```GDScript
peer_a.update_visibility(peer_b)
peer_b.update_visibility(peer_b)
```
Then save that value _somewhere_, which will need some sort of indexing (a map), and read it back from your callback. Its really inefficient and cumbersome IMO and it will not scale as well.
```

**Comment by jgillich** [comment]
> AFAIK this logic only works server side, client has no authority. I am not entirely sure why you think the peer is not enough, you can obtain the node from it.

Correct me if I'm wrong, but the client needs authority over its MultiplayerSynchronizers so it's allowed to push the values to the server. That's also how it's done in the [bomber example](https://github.com/godotengine/godot-demo-projects/blob/044afe7976735c628045759b3575b7175062535f/networking/multiplayer_bomber/player.gd#L21). Therefore, the server can't update visibility for clients. Would love to be wrong on this though :smile: (some docs on the new sync feature would be sweet, right now I'm mostly guessing based on examples and some of the C++ code that I don't fully understand)

**Comment by reduz** [comment]
@jgillich In practice, Synchronizer is only for Server -> Peer, not the other way around. Ideally you want to do the client -> server communication via RPC so the server can validate what you are trying to do.

**Comment by Faless** [comment]
I updated the PR with the simple visibility functions, and the callable as override.

The examples above becomes:

Much simpler for the simple case.

```gdscript
# Callback from Area node
func _player_visibility_enter(body):
    # Assuming body name is player id.
    set_peer_visibility(str(body.name).to_int(), true)

# Other area callback
func _player_visibility_exit(body):
    # Assuming body name is player id.
    set_peer_visibility(str(body.name).to_int(), false)
```

Pretty much the same for the complex scenario with a small change in callback registration.
```gdscript
# limited_sync.gd

func _enter_tree():
    # You can also set the override from a parent node.
    set_visibility_override(NetVisibility.is_visible_to.bind(self))
```

**Comment by Faless** [comment]
> Correct me if I'm wrong, but the client needs authority over its MultiplayerSynchronizers so it's allowed to push the values to the server. That's also how it's done in the [bomber example](https://github.com/godotengine/godot-demo-projects/blob/044afe7976735c628045759b3575b7175062535f/networking/multiplayer_bomber/player.gd#L21).

A note on this, which I forgot to answer.
The example might indeed be misleading in this sense, but it was mostly shown as a possible configuration, not necessarily the best solution in most cases (and not even in that particular case to be fair). It just happens that the bomber demo has a fully disclosed network state and there is no need to limit peer visibility (i.e. everyone seeing everyone else is part of the game), so it was okay (and fast) to sync that way.

While you can still use replication to sync the "keyboard state" from the client to the server only (via the `set_peer_visibility(1, true)` function, or returning `true` for `1` in the callback), it's probably not worth the extra effort over RPCs, which also give you greater control.

I was also thinking about changing the bomber demo to make visibility matter, but it's also already quite overloaded of information, so we might need rethinking, or to be split it up as part of a tutorial of some kind.

**Comment by Faless** [comment]
Okay, I've updated the PR so that callbacks behaves like filters (and updated the demos/snippet):

```gdscript
func _ready():
	$sync.add_visibility_filter(custom_visibility)

func custom_visibility(p_peer: int) -> bool:
	return true # Custom logic here

func block_filter(peer) -> bool: 
	return false # Always blocked

func _on_item_toggled(p_pressed):
	if p_pressed:
		# Block visibility from all
		$sync.add_visibility_filter(block_filter)
	else:
		# Remove visibility block
		$sync.remove_visibility_filter(block_filter)
```

**Review by reduz — APPROVED** [review]

File: core/multiplayer/multiplayer_api.h [hunk 1]
```diff
@@ -77,7 +77,8 @@ class MultiplayerCacheInterface : public RefCounted {
 	virtual void process_confirm_path(int p_from, const uint8_t *p_packet, int p_packet_len) {}
 
 	// Returns true if all peers have cached path.
-	virtual bool send_object_cache(Object *p_obj, NodePath p_path, int p_target, int &p_id) { return false; }
+	virtual bool send_object_cache(Object *p_obj, int p_target, int &p_id) { return false; }
```
Comment (akien-mga) [inline core/multiplayer/multiplayer_api.h:80]:
```suggestion
	virtual bool send_object_cache(Object *p_obj, int p_target, int &r_id) { return false; }
```

File: core/multiplayer/multiplayer_api.h [hunk 2]
```diff
@@ -160,7 +161,8 @@ class MultiplayerAPI : public RefCounted {
 	Error replication_start(Object *p_object, Variant p_config);
 	Error replication_stop(Object *p_object, Variant p_config);
 	// Cache API
-	bool send_object_cache(Object *p_obj, NodePath p_path, int p_target, int &p_id);
+	bool send_object_cache(Object *p_obj, int p_target, int &p_id);
```
Comment (akien-mga) [inline core/multiplayer/multiplayer_api.h:164]: Same here.

File: doc/classes/MultiplayerSynchronizer.xml [hunk 3]
```diff
@@ -6,12 +6,64 @@
 	</description>
 	<tutorials>
 	</tutorials>
+	<methods>
+		<method name="add_visibility_filter">
+			<return type="void" />
+			<argument index="0" name="filter" type="Callable" />
+			<description>
+			</description>
```
Comment (akien-mga) [inline doc/classes/MultiplayerSynchronizer.xml:14]: Would be great if you could write the docs too.

File: scene/multiplayer/multiplayer_spawner.cpp [hunk 4]
```diff
@@ -71,7 +71,7 @@ bool MultiplayerSpawner::_get(const StringName &p_name, Variant &r_ret) const {
 }
 
 void MultiplayerSpawner::_get_property_list(List<PropertyInfo> *p_list) const {
-	p_list->push_back(PropertyInfo(Variant::INT, "_spawnable_scene_count", PROPERTY_HINT_NONE, "", PROPERTY_USAGE_EDITOR | PROPERTY_USAGE_ARRAY, "Scenes,scenes/"));
+	p_list->push_back(PropertyInfo(Variant::INT, "_spawnable_scene_count", PROPERTY_HINT_NONE, "", PROPERTY_USAGE_EDITOR | PROPERTY_USAGE_ARRAY, "Auto Spawn List,scenes/"));
```
Comment (akien-mga) [inline scene/multiplayer/multiplayer_spawner.cpp:74]: Curious to see if that breaks GDExtension too (https://github.com/godotengine/godot/issues/63066).

File: doc/classes/MultiplayerSynchronizer.xml (hunk 3 above)
Comment (akien-mga) [reply doc/classes/MultiplayerSynchronizer.xml:14]: As discussed this will be done in another PR.

**Review by akien-mga — APPROVED** [review]
